In [1]:
import pandas as pd

In [170]:
final = pd.read_csv('./Data/sample_sub_reg.csv')

In [171]:
final

Id    SalePrice
0       2  181479.1217
1       4  181479.1217
2       6  181479.1217
3       7  181479.1217
4      17  181479.1217
5      18  181479.1217
6      22  181479.1217
7      27  181479.1217
8      31  181479.1217
9      36  181479.1217
10     39  181479.1217
11     40  181479.1217
12     42  181479.1217
13     46  181479.1217
14     53  181479.1217
15     58  181479.1217
16     61  181479.1217
17     72  181479.1217
18     79  181479.1217
19     80  181479.1217
20     82  181479.1217
21     83  181479.1217
22     85  181479.1217
23     88  181479.1217
24     92  181479.1217
25     93  181479.1217
26     95  181479.1217
27    102  181479.1217
28    103  181479.1217
29    106  181479.1217
..    ...          ...
849  2825  181479.1217
850  2828  181479.1217
851  2831  181479.1217
852  2843  181479.1217
853  2845  181479.1217
854  2847  181479.1217
855  2855  181479.1217
856  2856  181479.1217
857  2859  181479.1217
858  2865  181479.1217
859  2867  181479.1217
860  2869  181479.1217
861  2870  181479.1217
862  2874  181479.1217
863  2876  181479.1217
864  2889  181479.1217
865  2892  181479.1217
866  2895  181479.1217
867  2898  181479.1217
868  2901  181479.1217
869  2902  181479.1217
870  2908  181479.1217
871  2909  181479.1217
872  2916  181479.1217
873  2917  181479.1217
874  2919  181479.1217
875  2921  181479.1217
876  2922  181479.1217
877  2925  181479.1217
878  2928  181479.1217

[879 rows x 2 columns]

In [115]:
train = pd.read_csv('./Data/train.csv')

In [116]:
#train.info()

In [117]:
train['Fence'].fillna(0, inplace = True)
train['Misc Feature'].fillna(0, inplace= True)
train['Alley'].fillna(0,inplace=True)

In [118]:
#Mostly NaN. Killed them all, god will sort them out. He dosen't make mistakes

train.drop(['Pool QC','Fireplace Qu', 'Garage Yr Blt'],axis = 1, inplace=True)

In [119]:
#train.info()

In [120]:
train.head()

Id        PID  MS SubClass MS Zoning  Lot Frontage  Lot Area Street Alley  \
0  109  533352170           60        RL           NaN     13517   Pave     0   
1  544  531379050           60        RL          43.0     11492   Pave     0   
2  153  535304180           20        RL          68.0      7922   Pave     0   
3  318  916386060           60        RL          73.0      9802   Pave     0   
4  255  906425045           50        RL          82.0     14235   Pave     0   

  Lot Shape Land Contour    ...     3Ssn Porch Screen Porch Pool Area Fence  \
0       IR1          Lvl    ...              0            0         0     0   
1       IR1          Lvl    ...              0            0         0     0   
2       Reg          Lvl    ...              0            0         0     0   
3       Reg          Lvl    ...              0            0         0     0   
4       IR1          Lvl    ...              0            0         0     0   

  Misc Feature Misc Val Mo Sold Yr Sold  Sale Type  SalePrice  
0            0        0       3    2010        WD      130500  
1            0        0       4    2009        WD      220000  
2            0        0       1    2010        WD      109000  
3            0        0       4    2010        WD      174000  
4            0        0       3    2010        WD      138500  

[5 rows x 78 columns]

In [121]:
#train#.drop(['Unnamed: 0'], axis=1, inplace=True)

In [122]:

train.columns = [i.lower().replace(' ','_') for i in train.columns.get_values()]


In [123]:
train = train[train.lot_frontage.isna() == False]

train = train[train.mas_vnr_type.isna() == False]

train = train[train.bsmtfin_sf_1.isna() == False]

train = train[train.bsmt_full_bath.isna() == False]

train = train[train.garage_cars.isna() == False]

In [124]:
train.bsmt_qual.fillna(0, inplace=True)
train.bsmt_cond.fillna(0, inplace=True)
train.bsmt_exposure.fillna(0,inplace=True)
train.bsmtfin_type_1.fillna(0,inplace=True)
train.bsmtfin_type_2.fillna(0, inplace=True)
train.garage_type.fillna(0, inplace=True)
train.garage_finish.fillna(0, inplace=True)
train.garage_qual.fillna(0, inplace=True)
train.garage_cond.fillna(0, inplace=True)

In [125]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1702 entries, 1 to 2050
Data columns (total 78 columns):
id                 1702 non-null int64
pid                1702 non-null int64
ms_subclass        1702 non-null int64
ms_zoning          1702 non-null object
lot_frontage       1702 non-null float64
lot_area           1702 non-null int64
street             1702 non-null object
alley              1702 non-null object
lot_shape          1702 non-null object
land_contour       1702 non-null object
utilities          1702 non-null object
lot_config         1702 non-null object
land_slope         1702 non-null object
neighborhood       1702 non-null object
condition_1        1702 non-null object
condition_2        1702 non-null object
bldg_type          1702 non-null object
house_style        1702 non-null object
overall_qual       1702 non-null int64
overall_cond       1702 non-null int64
year_built         1702 non-null int64
year_remod/add     1702 non-null int64
roof_style         1

In [126]:
train.to_csv('./Data/train_noBV.csv', index=False) #Let's not overwrite this

In [127]:
train = pd.read_csv('./Data/train_noBV.csv')

In [128]:
train.head()

id        pid  ms_subclass ms_zoning  lot_frontage  lot_area street alley  \
0  544  531379050           60        RL          43.0     11492   Pave     0   
1  153  535304180           20        RL          68.0      7922   Pave     0   
2  318  916386060           60        RL          73.0      9802   Pave     0   
3  255  906425045           50        RL          82.0     14235   Pave     0   
4  138  535126040           20        RL         137.0     16492   Pave     0   

  lot_shape land_contour    ...     3ssn_porch screen_porch pool_area fence  \
0       IR1          Lvl    ...              0            0         0     0   
1       Reg          Lvl    ...              0            0         0     0   
2       Reg          Lvl    ...              0            0         0     0   
3       IR1          Lvl    ...              0            0         0     0   
4       IR1          Lvl    ...              0            0         0     0   

  misc_feature misc_val mo_sold yr_sold  sale_type  saleprice  
0            0        0       4    2009        WD      220000  
1            0        0       1    2010        WD      109000  
2            0        0       4    2010        WD      174000  
3            0        0       3    2010        WD      138500  
4            0        0       6    2010        WD      190000  

[5 rows x 78 columns]

In [129]:
#train.info()

In [130]:
train = pd.get_dummies(train, columns=['ms_zoning','street','alley','lot_shape','land_contour','utilities','lot_config', 
                               'land_slope','neighborhood','condition_1','condition_2','bldg_type','house_style',
                               'roof_style','roof_matl','exterior_1st','exterior_2nd','mas_vnr_type','exter_qual',
                               'exter_cond', 'foundation','bsmt_qual','bsmt_cond','bsmt_exposure','bsmtfin_type_1',
                               'bsmtfin_type_2','heating','heating_qc','central_air','electrical','kitchen_qual',
                               'functional','garage_type','garage_finish','garage_qual','garage_cond','paved_drive',
                               'fence','misc_feature','sale_type'])

In [169]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1702 entries, 0 to 1701
Columns: 286 entries, ms_subclass to sale_type_WD 
dtypes: float64(10), int64(25), uint8(251)
memory usage: 882.7 KB


In [133]:
train.shape

(1702, 289)

In [134]:
y_train = pd.DataFrame()
y_train['saleprice'] = train.saleprice

In [135]:
y_train.to_csv('./Data/y_train.csv', index=False)

In [136]:
from sklearn.utils import column_or_1d

In [137]:
y_train_array = column_or_1d(y_train)

In [138]:
X_train = pd.DataFrame()

In [139]:
X_train = train

In [140]:
X_train.drop(['id', 'pid','saleprice'],axis=1 ,inplace=True)

In [141]:
X_train.shape

(1702, 286)

In [63]:
X_train.to_csv('./Data/X_train.csv')

286 columns is a lot. Lets use Lasso to trim that down.

In [64]:
from sklearn.linear_model import Lasso, LassoCV

In [66]:
import numpy as np

In [83]:
l_alphas = np.logspace(-4, -.5, 30)
lasso_model = LassoCV(alphas=l_alphas, cv=5)
lasso_model = lasso_model.fit(X_train, y_train_array)

/home/steven/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)
/home/steven/anaconda3/lib/python3.6/site-packages/sklearn/linear_model/coordinate_descent.py:491: ConvergenceWarning: Objective did not converge. You might want to increase the number of iterations. Fitting data with very small alpha may cause precision problems.
  ConvergenceWarning)


In [84]:
lasso_optimal_alpha =lasso_model.alpha_

In [85]:
lasso_optimal_alpha

0.31622776601683794

In [154]:
X_test = pd.read_csv('./Data/test.csv')

In [155]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 879 entries, 0 to 878
Data columns (total 80 columns):
Id                 879 non-null int64
PID                879 non-null int64
MS SubClass        879 non-null int64
MS Zoning          879 non-null object
Lot Frontage       719 non-null float64
Lot Area           879 non-null int64
Street             879 non-null object
Alley              58 non-null object
Lot Shape          879 non-null object
Land Contour       879 non-null object
Utilities          879 non-null object
Lot Config         879 non-null object
Land Slope         879 non-null object
Neighborhood       879 non-null object
Condition 1        879 non-null object
Condition 2        879 non-null object
Bldg Type          879 non-null object
House Style        879 non-null object
Overall Qual       879 non-null int64
Overall Cond       879 non-null int64
Year Built         879 non-null int64
Year Remod/Add     879 non-null int64
Roof Style         879 non-null object
Roof M

In [157]:
X_test.columns = [i.lower().replace(' ','_') for i in X_test.columns.get_values()]

In [158]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 879 entries, 0 to 878
Data columns (total 80 columns):
id                 879 non-null int64
pid                879 non-null int64
ms_subclass        879 non-null int64
ms_zoning          879 non-null object
lot_frontage       719 non-null float64
lot_area           879 non-null int64
street             879 non-null object
alley              58 non-null object
lot_shape          879 non-null object
land_contour       879 non-null object
utilities          879 non-null object
lot_config         879 non-null object
land_slope         879 non-null object
neighborhood       879 non-null object
condition_1        879 non-null object
condition_2        879 non-null object
bldg_type          879 non-null object
house_style        879 non-null object
overall_qual       879 non-null int64
overall_cond       879 non-null int64
year_built         879 non-null int64
year_remod/add     879 non-null int64
roof_style         879 non-null object
roof_m

In [159]:
X_test.drop(['pool_qc','fireplace_qu', 'garage_yr_blt'],axis = 1, inplace=True)

In [160]:
X_test.fence.fillna(0, inplace = True)
X_test.misc_feature.fillna(0, inplace= True)
X_test.alley.fillna(0,inplace=True)
X_test.bsmt_qual.fillna(0, inplace=True)
X_test.bsmt_cond.fillna(0, inplace=True)
X_test.bsmt_exposure.fillna(0,inplace=True)
X_test.bsmtfin_type_1.fillna(0,inplace=True)
X_test.bsmtfin_type_2.fillna(0, inplace=True)
X_test.garage_type.fillna(0, inplace=True)
X_test.garage_finish.fillna(0, inplace=True)
X_test.garage_qual.fillna(0, inplace=True)
X_test.garage_cond.fillna(0, inplace=True)

In [163]:
X_test = X_test[X_test.lot_frontage.isna() == False]

X_test = X_test[X_test.mas_vnr_type.isna() == False]

X_test = X_test[X_test.bsmtfin_sf_1.isna() == False]

X_test = X_test[X_test.bsmt_full_bath.isna() == False]

X_test = X_test[X_test.garage_cars.isna() == False]

X_test = X_test[X_test.electrical.isna() == False]

In [164]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 717 entries, 0 to 878
Data columns (total 77 columns):
id                 717 non-null int64
pid                717 non-null int64
ms_subclass        717 non-null int64
ms_zoning          717 non-null object
lot_frontage       717 non-null float64
lot_area           717 non-null int64
street             717 non-null object
alley              717 non-null object
lot_shape          717 non-null object
land_contour       717 non-null object
utilities          717 non-null object
lot_config         717 non-null object
land_slope         717 non-null object
neighborhood       717 non-null object
condition_1        717 non-null object
condition_2        717 non-null object
bldg_type          717 non-null object
house_style        717 non-null object
overall_qual       717 non-null int64
overall_cond       717 non-null int64
year_built         717 non-null int64
year_remod/add     717 non-null int64
roof_style         717 non-null object
roof_

In [165]:
X_test = pd.get_dummies(X_test, columns=['ms_zoning','street','alley','lot_shape','land_contour','utilities','lot_config', 
                               'land_slope','neighborhood','condition_1','condition_2','bldg_type','house_style',
                               'roof_style','roof_matl','exterior_1st','exterior_2nd','mas_vnr_type','exter_qual',
                               'exter_cond', 'foundation','bsmt_qual','bsmt_cond','bsmt_exposure','bsmtfin_type_1',
                               'bsmtfin_type_2','heating','heating_qc','central_air','electrical','kitchen_qual',
                               'functional','garage_type','garage_finish','garage_qual','garage_cond','paved_drive',
                               'fence','misc_feature','sale_type'])

In [166]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 717 entries, 0 to 878
Columns: 272 entries, id to sale_type_WD 
dtypes: float64(10), int64(27), uint8(235)
memory usage: 377.4 KB


In [167]:
X_test.drop(['id', 'pid'],axis=1 ,inplace=True)

In [168]:
X_test.shape

(717, 270)

In [104]:
cleo = lasso_model.predict(X_test)

ValueError: shapes (718,270) and (286,) not aligned: 270 (dim 1) != 286 (dim 0)